### PR TITLE
Lidarprofile ngeo22

### PR DESCRIPTION
**Pull request by monodo**
TO BE DONE BEFORE MERING

- [x] Setup type for [d3.linearScale](https://github.com/monodo/ngeo/blob/lidarprofile_ngeo22/contribs/gmf/src/services/lidarprofileconfig.js#L33-L44)
- [ ] Fix altitude decoding issue [here](https://github.com/monodo/ngeo/blob/lidarprofile_ngeo22/contribs/gmf/src/d3-lidar-profile/lidarprofileLoader.js#L360)
- [x] Setup type for Pytree service response [here](https://github.com/monodo/ngeo/blob/lidarprofile_ngeo22/contribs/gmf/src/services/lidarprofileconfig.js#L86)
- [x] Move the [scaleX, scaleY and previousDomainX](https://github.com/monodo/ngeo/blob/lidarprofile_ngeo22/contribs/gmf/src/services/lidarprofileconfig.js#L33-L44) to a more pertinent place
- [x] Merge latest 2.2 branch into this one

Live demo [here](https://sitn.ne.ch/dev_internet2/desktop) 

@ger-benjamin please review 

---

Edited by ger-benjamin
New TODO:
 1. [ ] fix "FIXME missing description"
 1. [x] Create an example
 1. [x] Move the application example in the desktop_alt
 1. [x] Standardize name of pytree api (looks strange: `/profile_config_gmf2` looks strange no ? https://github.com/monodo/ngeo/blob/d2e8cb8a5496610eec93bbe5faef856e5ed9fd95/contribs/gmf/src/d3-lidar-profile/lidarprofileConfig.js#L65)
 1. [x] rework download functions in https://github.com/monodo/ngeo/blob/d2e8cb8a5496610eec93bbe5faef856e5ed9fd95/contribs/gmf/src/d3-lidar-profile/lidarprofileUtils.js
 1. [x] Add translation on all texts

Do all we can on available time to:
 - Improve general architecture
- Simplify functions when it's possible

---

Debug:

- [x] easy: Cleanall function (in lidarpanel controller) works not completely (line persist on the map)
- [x] easy: Fix some error that appears in the console
- [ ] not easy: Can't drag and drop profile to move along the line
- [ ] not easy: It's hard to do a measure
- [x] not easy: profile is compiled but doesn't work (first step: add missing require !)
- [ ] easy-but-with-server-part: rework suspicious code part like: https://github.com/camptocamp/ngeo/blob/b212b7fe4b03d697a08cbcff42153507b8b1729e/contribs/gmf/src/d3-lidar-profile/lidarprofilePlot.js#L90-L94

### CODE REVIEW

**Comment by ger-benjamin** [comment]
First: Thanks for this huge and good work !

Then, some comment :-p
- Can you please rebase to get less commit. For instance, a "wip" commit is never interesting, you can merge into the previous one. If we can have maximum 10 commit, it's better (you can merge all commit in one if you want.) (`git rebase -i HEAD~63`, then replace `pick` by `f` to meld a commit into the previous one)
- Can you remove the lidarprofile from the desktop and add it in the desktop_alt ? Then also create a example with it (look like it's done for the contribs/gmf/examples/profile example). We don't want the lidarprofile directly in the desktop example, because we don't have internal data to use it.
- (I'll continue soon...)

File: contribs/gmf/apps/desktop/js/controller.js [hunk 1]
```diff
@@ -31,9 +31,9 @@ goog.require('ngeo.proj.EPSG21781');
  */
 app.DesktopController = function($scope, $injector) {
   gmf.AbstractDesktopController.call(this, {
-    srid: 21781,
+    srid: 2056,
     mapViewConfig: {
-      center: [632464, 185457],
+      center: [2551894, 1202362],
```
Comment (ger-benjamin) [inline contribs/gmf/apps/desktop/js/controller.js]: To revert before to merge (both lines)

File: contribs/gmf/less/lidarprofile.less [hunk 2]
```diff
@@ -0,0 +1,58 @@
+@lidar-profile-height: 35rem;
+.gmf-lidar-profile-container {
+  position: relative;
+  overflow: hidden;
+  @profile-legend-width: 15rem;
```
Comment (ger-benjamin) [inline contribs/gmf/less/lidarprofile.less]: Strange notation here, what that means ?

File: contribs/gmf/options/gmfx.js [hunk 3]
```diff
@@ -112,6 +112,143 @@ gmfx.GridSource.prototype.source;
...
+//  * defaultColor: (string|undefined),
+//  * pointClouds: (Object|undefined),
+//  * pointAttributesRaw: (gmfx.lidarPointAttributeList),
+//  * defaultAttribute: (gmfx.lidarPointAttribute),
+//  * defaultPointAttribute: (string|undefined),
+//  * debug: (boolean|undefined),
+//  * pytreeLinestring: (string|undefined)
+//  * }}
+//  */
+// gmfx.LidarProfileConfig;
```
Comment (ger-benjamin) [inline contribs/gmf/options/gmfx.js]: To remove ?

File: contribs/gmf/src/d3-lidar-profile/lidarprofile.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
+goog.provide('gmf.lidarProfile');
+
+/**
+* Provides a D3js component to be used to draw an lidar point cloud
+* profile chart.
+* Requires access to a Pytree webservice: https://github.com/sitn/pytree
+*
+* @constructor
+* @param {Object} options from pytree
```
Comment (ger-benjamin) [inline contribs/gmf/src/d3-lidar-profile/lidarprofile.js]: It's possible to provide a better type ?

File: contribs/gmf/src/d3-lidar-profile/lidarprofile.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+*/
+
+gmf.lidarProfile = function(options) {
+
+  /**
+  * The instance of gmf.lidarProfile class used to access loader instance
+  * properties and methods from plot and measure instances
+  * @type {gmf.lidarProfile}
+  */
+  const parent = this;
```
Comment (ger-benjamin) [inline contribs/gmf/src/d3-lidar-profile/lidarprofile.js]: Can't you remove this constant and use directly `this` in constructor's call below ?

File: contribs/gmf/src/d3-lidar-profile/lidarprofileMeasure.js [hunk 6]
```diff
@@ -0,0 +1,200 @@
+goog.provide('gmf.lidarProfile.measure');
+
+
+/**
+* Measure tool for the d3 chart
+* @constructor
+* @param {Object} options to be defined in gmfx
```
Comment (ger-benjamin) [inline contribs/gmf/src/d3-lidar-profile/lidarprofileMeasure.js]: to be defined

File: contribs/gmf/src/d3-lidar-profile/lidarprofileMeasure.js [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+      intensity: undefined,
+      classification: undefined,
+      set: false
+    }
+  };
+
+  /**
+  * @type {gmf.lidarProfile}
+  */
+  this.parent_ = parent;
```
Comment (ger-benjamin) [inline contribs/gmf/src/d3-lidar-profile/lidarprofileMeasure.js]: Missing `@private`

File: contribs/gmf/src/d3-lidar-profile/lidarprofileMeasure.js [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+  svg.selectAll('#text_m').remove();
+  svg.selectAll('#start_m').remove();
+  svg.selectAll('#end_m').remove();
+  svg.selectAll('#line_m').remove();
+
+  d3.select('#height_measure').html('');
+  d3.select('svg#profileSVG').on('click', null);
+
+  d3.select('svg#profileSVG').style('cursor', 'default');
+
```
Comment (ger-benjamin) [inline contribs/gmf/src/d3-lidar-profile/lidarprofileMeasure.js]: (Please, remove too many useless spaces in each files)

File: contribs/gmf/src/d3-lidar-profile/lidarprofileMeasure.js [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+* Activate the measure tool
+* @export
+*/
+gmf.lidarProfile.measure.prototype.setMeasureActive = function() {
+
+  const that = this;
+
+  d3.select('svg#profileSVG').style('cursor', 'pointer');
+
+  function measureHeigt() {
```
Comment (ger-benjamin) [inline contribs/gmf/src/d3-lidar-profile/lidarprofileMeasure.js]: Please, move out this function (to `gmf.lidarProfile.measure.prototype.measureHeight`)

And please never do thing like `const that = this;`, use my_function.bind(this) and use `this` inside

File: contribs/gmf/src/d3-lidar-profile/lidarprofilePlot.js [hunk 10]
```diff
@@ -0,0 +1,320 @@
+goog.provide('gmf.lidarProfile.plot');
+
+/**
+* @constructor
+* @param {Object} options to be defined in gmfx
```
Comment (ger-benjamin) [inline contribs/gmf/src/d3-lidar-profile/lidarprofilePlot.js]: to be defined

File: contribs/gmf/src/d3-lidar-profile/lidarprofilePlot.js [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+
+/**
+ * Draw the points to the canvas element
+ * @param {gmfx.LidarProfilePoints} points of the profile
+ * @param {string} material used to determine point color
+ * @export
+*/
+gmf.lidarProfile.plot.prototype.drawPoints = function(points, material) {
+  let i = -1;
+  const n = points.distance.length;
```
Comment (ger-benjamin) [inline contribs/gmf/src/d3-lidar-profile/lidarprofilePlot.js]: Please, use human-readable name for variables (except for `i`)

File: contribs/gmf/src/d3-lidar-profile/lidarprofilePlot.js [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+      .domain([
+        domainHeightCentroid - domainScaledHeight / 2,
+        domainHeightCentroid + domainScaledHeight / 2])
+      .range([height, 0]);
+  }
+
+  this.options.profileConfig.scaleX = sx;
+  this.options.profileConfig.scaleY = sy;
+
+  const that = this;
```
Comment (ger-benjamin) [inline contribs/gmf/src/d3-lidar-profile/lidarprofilePlot.js]: Remove that please

File: contribs/gmf/src/d3-lidar-profile/lidarprofilePlot.js [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+    svg.selectAll('#highlightCircle').remove();
+
+    d3.select('svg#profileSVG').append('circle')
+      .attr('id', 'highlightCircle')
+      .attr('cx', cx)
+      .attr('cy', cy)
+      .attr('r', pointSize + 1)
+      .style('fill', 'orange');
+
+    const html = `Distance: ${Math.round(10 * p.distance) / 10}<br>
```
Comment (ger-benjamin) [inline contribs/gmf/src/d3-lidar-profile/lidarprofilePlot.js]: I don't know exactly how to do this now, but can you start to move this into an function (getTooltipHtml or something like that ?). We must be able to translate texts inside.

File: contribs/gmf/src/d3-lidar-profile/lidarprofilePlot.js [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+
+  this.options.profileConfig.previousDomainX = sx.domain();
+  this.options.profileConfig.previousDomainY = sy.domain();
+
+};
+
+/**
+ * Update the Openlayers overlay that displays point position and attributes values
+ * @export
+ * @param {gmf.lidarProfile.plot} that scope of the plot class
```
Comment (ger-benjamin) [inline contribs/gmf/src/d3-lidar-profile/lidarprofilePlot.js]: Can't be a useful parameter, please remove it and use "this" inside the function

File: contribs/gmf/src/d3-lidar-profile/lidarprofileUtils.js [hunk 15]
```diff
@@ -0,0 +1,487 @@
+goog.provide('gmf.lidarProfile.utils');
+
+
+/**
+* @constructor
+* @param {Object} options to be defined in gmfx
```
Comment (ger-benjamin) [inline contribs/gmf/src/d3-lidar-profile/lidarprofileUtils.js]: TBD

File: contribs/gmf/src/d3-lidar-profile/lidarprofileUtils.js [hunk 16]
```diff
@@ -0,0 +1,487 @@
...
+
+/**
+* Create a fake link and make the browser download the referenced URL
+* @param {string} filename csv file name
+* @param {string} dataUrl fake url from which to download the csv file
+* @export
+*/
+gmf.lidarProfile.utils.prototype.downloadDataUrlFromJavascript = function(filename, dataUrl) {
+
+  const link = document.createElement('a');
```
Comment (ger-benjamin) [inline contribs/gmf/src/d3-lidar-profile/lidarprofileUtils.js]: you should be able to use The saveAs function from FileSaver library, see: https://github.com/camptocamp/ngeo/blob/2.2/src/services/download.js and https://github.com/camptocamp/ngeo/blob/2.2/src/services/featurehelper.js#L844-L846

File: contribs/gmf/src/d3-lidar-profile/lidarprofileUtils.js [hunk 17]
```diff
@@ -0,0 +1,487 @@
...
+    }
+  }
+  return minVal;
+};
+
+/**
+* Create an UUID unique identifier
+* @return {string} uuid
+*/
+gmf.lidarProfile.utils.prototype.UUID = function() {
```
Comment (ger-benjamin) [inline contribs/gmf/src/d3-lidar-profile/lidarprofileUtils.js]: I'm pretty sure we have a tool for that. @fredj  Do you have an idea ?

File: contribs/gmf/src/directives/lidarpanel.js [hunk 18]
```diff
@@ -0,0 +1,349 @@
+goog.provide('gmf.lidarPanelComponent');
+
+goog.require('gmf');
+goog.require('gmf.lidarProfile');
+/**
+* @suppress {extraRequire}
```
Comment (ger-benjamin) [inline contribs/gmf/src/directives/lidarpanel.js]: Can you write it on one line (not 3) please ?

Comment (ger-benjamin) [inline contribs/gmf/src/directives/lidarpanel.js]: Cane you write it on 1 line (not 3) please

File: contribs/gmf/src/directives/lidarprofile.js [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+ * @return {string} Template URL.
+ * @ngInject
+ */
+function gmfLidarProfileTemplateUrl($element, $attrs, gmfLidarProfileTemplateUrl) {
+  return gmfLidarProfileTemplateUrl($element, $attrs);
+}
+
+
+/**
+ * Provide a component that display a lidar profile panel.
```
Comment (ger-benjamin) [inline contribs/gmf/src/directives/lidarprofile.js]: Please, improve the documentation. See: https://github.com/camptocamp/ngeo/blob/2.2/contribs/gmf/src/directives/displayquerygrid.js#L59-L79

(Example and htmlattributes)

File: contribs/gmf/src/directives/lidarpanel.js [hunk 20]
```diff
@@ -0,0 +1,349 @@
...
+    });
+
+};
+
+
+/**
+ * @private
+ */
+gmf.LidarPanelController.prototype.$onInit = function() {
+  this.line = this.line;
```
Comment (ger-benjamin) [inline contribs/gmf/src/directives/lidarpanel.js]: Is this useful ?

File: contribs/gmf/src/directives/lidarprofile.js [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+ * @ngdoc component
+ * @ngname gmfLidarProfile
+ */
+gmf.lidarProfileComponent = {
+  controller: 'GmfLidarProfileController',
+  bindings: {
+    'active': '=gmfLidarProfileActive',
+    'line': '=gmfLidarProfileLine',
+    'getMapFn': '&?gmfLidarProfileMap',
```
Comment (ger-benjamin) [inline contribs/gmf/src/directives/lidarprofile.js]: That's not false, but that's better to use: `<?gmfLidarProfileMap` here and: `gmf-lidar-profile-map=::ctrl.map` (with the double `::` in the html)...

File: contribs/gmf/src/directives/lidarprofile.js [hunk 22]
```diff
@@ -0,0 +1,137 @@
...
+    });
+
+};
+
+
+/**
+ * @private
+ */
+gmf.LidarProfileController.prototype.$onInit = function() {
+  this.map_ = this['getMapFn'] ? this['getMapFn']() : null;
```
Comment (ger-benjamin) [inline contribs/gmf/src/directives/lidarprofile.js]: ...then, with the `<` binding, not the `&`, you will be able to remove this notation

File: contribs/gmf/src/directives/partials/lidarpanel.html [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+  <hr>
+  Classes:
+  <br></br>
+  <div ng-repeat="(key, classification) in ::$ctrl.getClassification()">
+    <input ng-true-value="1" ng-false-value="0" ng-model="classification.visible" ng-change="$ctrl.setClassification(classification, key)" ng-checked="!!classification.visible" type="checkbox">
+    <span>{{classification.name}}</span>
+  </div>
+</div>
+<div ng-hide="$ctrl.ready">
+ <p>Initialisation en cours, merci de patienter...</p>
```
Comment (ger-benjamin) [inline contribs/gmf/src/directives/partials/lidarpanel.html]: All texts must be wrote: `<p translate>my texte<p>` or `<p>{{'my texte' | translate}}<p>` to be able to translate them

File: contribs/gmf/src/services/lidarprofileconfig.js [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+ */
+gmf.LidarProfileConfig = function($http, pytreeLidarProfileJsonUrl) {
+  /**
+   * @type {angular.$http}
+   * @private
+   */
+  this.$http_ = $http;
+
+  /**
+   * @type {string}
```
Comment (ger-benjamin) [inline contribs/gmf/src/services/lidarprofileconfig.js]: missing `@private`

File: contribs/gmf/src/services/lidarprofileconfig.js [hunk 25]
```diff
@@ -0,0 +1,119 @@
...
+   * @type {gmfx.lidarClassificationObjects}
+   */
+  this.profileConfig.classification;
+
+};
+
+/**
+* Initialize the service variables from Pytree profile_config_gmf2 route
+* @export
+* @return {Object} configuration values
```
Comment (ger-benjamin) [inline contribs/gmf/src/services/lidarprofileconfig.js]: It's possible to have a better type ?

File: contribs/gmf/options/gmfx.js (hunk 3 above)
Comment (monodo) [reply contribs/gmf/options/gmfx.js]: To be used for typing Pytree config service ouput. Help needed on this point...

File: contribs/gmf/src/d3-lidar-profile/lidarprofile.js (hunk 4 above)
Comment (monodo) [reply contribs/gmf/src/d3-lidar-profile/lidarprofile.js]: See previous comment ;-)

File: contribs/gmf/less/lidarprofile.less (hunk 2 above)
Comment (monodo) [reply contribs/gmf/less/lidarprofile.less]: rem: Relative to font-size of the root element. Used already in raster profile. More info [here](https://www.w3schools.com/cssref/css_units.asp)

Comment (ger-benjamin) [reply contribs/gmf/less/lidarprofile.less]: That's not the problem, what I didn't understand was the @profile-legend-width declared here. Now I understand a little bit but why here ? That looks strange

File: contribs/gmf/src/d3-lidar-profile/lidarprofile.js (hunk 5 above)
Comment (monodo) [reply contribs/gmf/src/d3-lidar-profile/lidarprofile.js]: Done

File: contribs/gmf/less/lidarprofile.less (hunk 2 above)
Comment (monodo) [reply contribs/gmf/less/lidarprofile.less]: Ok, @profile-legend-width moved to more logical position in less file

File: contribs/gmf/src/d3-lidar-profile/lidarprofileMeasure.js (hunk 7 above)
Comment (monodo) [reply contribs/gmf/src/d3-lidar-profile/lidarprofileMeasure.js]: Fixed

File: contribs/gmf/src/d3-lidar-profile/lidarprofilePlot.js (hunk 11 above)
Comment (monodo) [reply contribs/gmf/src/d3-lidar-profile/lidarprofilePlot.js]: Fixed

File: contribs/gmf/src/services/lidarprofileconfig.js (hunk 24 above)
Comment (monodo) [reply contribs/gmf/src/services/lidarprofileconfig.js]: Fixed

File: contribs/gmf/src/directives/partials/lidarpanel.html (hunk 23 above)
Comment (monodo) [reply contribs/gmf/src/directives/partials/lidarpanel.html]: Done

File: contribs/gmf/src/directives/lidarpanel.js (hunk 18 above)
Comment (monodo) [reply contribs/gmf/src/directives/lidarpanel.js]: Fixed

File: contribs/gmf/src/directives/lidarpanel.js (hunk 20 above)
Comment (monodo) [reply contribs/gmf/src/directives/lidarpanel.js]: No, removed

File: contribs/gmf/src/d3-lidar-profile/lidarprofile.js (hunk 5 above)
Comment (monodo) [reply contribs/gmf/src/d3-lidar-profile/lidarprofile.js]: Yes, fixed

File: contribs/gmf/less/lidarprofile.less (hunk 2 above)
Comment (monodo) [reply contribs/gmf/less/lidarprofile.less]: Fixed, moved higher in less file

File: contribs/gmf/src/d3-lidar-profile/lidarprofileMeasure.js (hunk 6 above)
Comment (monodo) [reply contribs/gmf/src/d3-lidar-profile/lidarprofileMeasure.js]: Idem, help for gmfx type definition needed

File: contribs/gmf/src/d3-lidar-profile/lidarprofileMeasure.js (hunk 7 above)
Comment (monodo) [reply contribs/gmf/src/d3-lidar-profile/lidarprofileMeasure.js]: Fixed

File: contribs/gmf/src/d3-lidar-profile/lidarprofileMeasure.js (hunk 8 above)
Comment (monodo) [reply contribs/gmf/src/d3-lidar-profile/lidarprofileMeasure.js]: Really ?

File: contribs/gmf/src/d3-lidar-profile/lidarprofilePlot.js (hunk 11 above)
Comment (monodo) [reply contribs/gmf/src/d3-lidar-profile/lidarprofilePlot.js]: Fixed

File: contribs/gmf/src/d3-lidar-profile/lidarprofileMeasure.js (hunk 9 above)
Comment (monodo) [reply contribs/gmf/src/d3-lidar-profile/lidarprofileMeasure.js]: Fixed

**Comment by monodo** [comment]
@ger-benjamin, I just removed all that = this stuff except for one in the loader. This latest commit breaks:
- The pan behaviour
- The PNG export (the svg element is no longer included)
- The build as d3.linearScale is not defined and closures does not accept direct call to an object

Could takt care of this

**Comment by ger-benjamin** [comment]
I'll have a try, but I can't spend much time on that this week.
I'll take this branch and add commits to fix this.

**Comment by monodo** [comment]
Here for instance: https://github.com/camptocamp/ngeo/pull/3464/commits/8cea4c6df0d78df7326b77f804ab0041c166fce9#diff-d5d5bc53c9e4d6f1b464c7cf75a0bdd4R190 

But there is still a that = this in the loader on the XHR request. As I though you would like to replace XHR by $http, I did not touch that part.

This last commits does brake the 3 elements listed, not in the build which is down due to false d3.linearScale typing.

**Comment by ger-benjamin** [comment]
Hi,

I've squashed commit into only 2 distinct commits. (If that's a problem for you, I can still rollback this merge because I've a fork here: https://github.com/ger-benjamin/ngeo/tree/lidarprofile_ngeo22)

I've done a lot of things. Now, the make check build and the code is more standard. But it take me 2 days to do that. That was really complicated because things are intricated in a complex way.  And to finish the TODO list (look at the first message), I think we need again 2 days :-/
I can't continue this week. But @kalbermattenm you can add you fix (you sent me by e-mail) if you want in this branch.

To know:
 - There is a new minor bug with the classification dropdown in the panel. (selected item is empty at first usage)
 - There is nothing more to cancel request. I've used the "ngeoDebounce" system to only query the pytree url after 200ms of a "break" in detected user's action.
 - I've moved the config in the d3-lidar-profile directory (now, I'm not sure that make sens, but that's not a big problem)

**Comment by monodo** [comment]
Thanks for the work done @ger-benjamin , looks good, except for the use of ngeoDebounce: it seems to me that this does not cancel requests that are "outdated" and thus will cause the client to draw wrong profiles on the canvas element. 200ms is not sufficient for deeper LODs and more will cause the UI to be not reactive enough.

File: contribs/gmf/src/d3-lidar-profile/lidarprofilePlot.js [hunk 26]
```diff
@@ -0,0 +1,386 @@
...
+      if (profileServerConfig.classification_colors[classification] &&
+            profileServerConfig.classification_colors[classification].visible) {
+
+        cx = this.scaleX(distance);
+        cy = this.scaleY(altitude);
+
+        ctx.beginPath();
+        ctx.moveTo(cx, cy);
+
+        if (material == 'COLOR_PACKED') {
```
Comment (ger-benjamin) [inline contribs/gmf/src/d3-lidar-profile/lidarprofilePlot.js]: this must always be fixed (use config from server, not a hard-coded thing)

**Comment by ger-benjamin** [comment]
> except for the use of ngeoDebounce: it seems to me that this does not cancel requests that are "outdated" and thus will cause the client to draw wrong profiles on the canvas element

The ngeoDebounce juste wait before to do some request. But then, if a request was done, the profile should (and will I think) ignore the previous request (complet or not), do a new request, clear the profile and display the new one.
Cancel the previous requests will not help. I think that was necessary with XMLHttpRequest(), but not with $http.  To test.

**Comment by kalbermattenm** [comment]
@ger-benjamin can I push the famous bugfix to this branch ?

**Comment by ger-benjamin** [comment]
Yes I leave it like this fornow

File: contribs/gmf/examples/lidarprofile.html [hunk 27]
```diff
@@ -0,0 +1,117 @@
...
+
+        .gmf-tooltip-measure {
+          font-weight: bold;
+        }
+    </style>
+  </head>
+  <body ng-controller="MainController as ctrl">
+    <div style="position: relative; float:left; margin-right:10px;">
+      <gmf-map gmf-map-map="ctrl.map"></gmf-map>
+      <p id="desc">This example shows how to use the <code>gmf-lidar-profile</code> with the <code>gmf-lidar-pranel</code>.</p>
```
Comment (kalbermattenm) [inline contribs/gmf/examples/lidarprofile.html]: There is a typo here: `gmf-lidar-pranel` should be `gmf-lidar-panel`

**Comment by kalbermattenm** [comment]
@ger-benjamin, could you tell me why I add to make ma last commit ? I do not get what happened here..?

Should I undo my last commit (https://github.com/camptocamp/ngeo/pull/3464/commits/33824c0fbff32d6d6fc832fe7f2fe6325e1b07ba) ? if yes, how ?

**Comment by kalbermattenm** [comment]
It seems related to your two last commits, and I do not get why the have the same comment, but the changes are obscure for me...

**Comment by ger-benjamin** [comment]
Strange. Do you have an uptodate branch ? (It seems but I don't have any other explication...)

**Comment by kalbermattenm** [comment]
Yeah, I installed the whole thing this morning...

**Comment by kalbermattenm** [comment]
#ger-benjamin, do you have any idea why the tests do not pass ?

**Comment by sbrunner** [comment]
I restart the Travis test and now it's OK :-)

**Comment by ger-benjamin** [comment]
@monodo @kalbermattenm 
I've updated a last time this PR.

Important note:
Without your last commit @kalbermattenm , It works and build. With your last commit, There is more interaction issues (no more hover on points, pan more difficult...). (It should build anymore)

Please address **all** the last not-checked checkbox of the first comment. I think you will be quicker than me. (I just take a look on the d3 pan zoom problems, and I give up quickly because I've no more time. But perhaps that can be useful: http://bl.ocks.org/sgruhier/1d692762f8328a2c9957)

Thanks.

**Comment by kalbermattenm** [comment]
I undid my changes and I have the same interaction issues as you are mentioning, thus I do not think that my commit is faulty.

Then I thought that testing it in a "online" demo would be another way to check it, but when I updated NGEO to use this branch and its last commit, then I get an empty panel:

image

@ger-benjamin, do you have an idea why I get this empty panel ?? (https://sitn.ne.ch/dev_internet2)

**Comment by kalbermattenm** [comment]
@ger-benjamin, in this commit https://github.com/camptocamp/ngeo/pull/3464/commits/c059eedb16d63f60a05ebe72bc158c2d0ef8bc20, why did you modify `contribs/gmf/src/controllers/abstractdesktop.js` ? It integrates the needed components, if you get rid of them, then we get the upper empty panel !

And I have the same bug as you described (interactions like pan and measure), but again I undid my changes and they are not the cause of these interactions bugs, check it here: https://sitn.ne.ch/dev_internet2/?debug)

File: contribs/gmf/less/desktop.less [hunk 28]
```diff
@@ -18,6 +18,7 @@
 @import 'contextualdata.less';
 @import 'fullscreenpopup.less';
 @import 'print.less';
+@import 'lidarprofile.less';
```
Comment (sbrunner) [inline contribs/gmf/less/desktop.less]: This should be in the file: contribs/gmf/apps/desktop_alt/less/main.less :-)

**Comment by ger-benjamin** [comment]
> why did you modify contribs/gmf/src/controllers/abstractdesktop.js ?

Because I've moved the lidar profile into the desktop_alt application. We won't have it in all desktop application, so I've added the requirement only in the desktop_alt controller. See:
https://github.com/camptocamp/ngeo/commit/c059eedb16d63f60a05ebe72bc158c2d0ef8bc20#diff-12c416dced7ccb963815cc118fd2a429R17

Do you have made a clean (cleanall ?) before your make serve ? it still work on my computer (except the interaction bug I've described here)

**Comment by kalbermattenm** [comment]
Yes, but sitn_c2cgeoportal loads the components through the abstract desktop controller, right ?

the desktop_alt and desktop stuff is here for the demos and not for applications (like sitn_c2cgeoportal) using ngeo, right ?

There is something I do not get...?

**Comment by ger-benjamin** [comment]
Oh sorry, I'll explain:
Your application inherit of the AbstractDesktopController (that inherit from the AbstractController). But we won't the LIDAR profile on each application. So you need to add the dependecy in your "final" controller, so in <project>/static-ngeo/js<application>.js

See: https://github.com/camptocamp/demo_geomapfish/blob/2.2/demo/static-ngeo/js/desktop.js (where we add the googstreeview for instance)

**Comment by kalbermattenm** [comment]
Hmmm.... So I think I got your point and added `goog.require('gmf.lidarPanelComponent');` to your desktop.js file... but then I do not get the result window unless I add `goog.require('gmf.lidarProfileComponent');` as well

**Comment by ger-benjamin** [comment]
Complement: 
 - all file in `src` and `/contribs/gmf/src` are imported in you final project (available in`node_module/ngeo`...)
 - All `examples`, `contribs/gmf/examples` and `contribs/gmf/apps` are NOT included in your final package.

I've added the requirement of the LIDAR panel in  `contribs/gmf/apps/desktop_alt/js/controller.js` So you have any file in your project that requirethe LIDAR profile. You must add the requirement in the app you want (so in `<project>/static-ngeo/js<application>.js`)

**Comment by kalbermattenm** [comment]
It is working now: https://sitn.ne.ch/dev_internet2/theme/main?debug

I undid my changes (https://sitn.ne.ch/dev_internet2/wsgi/node_modules/ngeo/contribs/gmf/src/d3-lidar-profile/lidarprofileManager.js), but still get the interaction issues.

I will check with @monodo if he has an idea...

**Comment by ger-benjamin** [comment]
Great. Now it works exactly like on my instance (without your commit).

 - With your commit, I don't know why but the point "hover" don't do anything :-/ (so that's not the case on your current example)
 - In all version, I've the zoom/pan problems (that's the case also on you current example)

**Comment by monodo** [comment]
Getting back to this...

A few bugs have been introduced recently [here](https://sitn.ne.ch/dev_internet2):
 - Each click on the canvas causes a full redraw of the profile. That was not the case and shouldbe fixed
- Switching from manual width to automatic width causes profile to redraw with classification color: it used to keep the active material, what seems more user friendly
- The pan behaviour is indeed broken. Most probably because the svg layout is not called on the correct pytree request....that's why the previous implementation assigned a uuid to each pytree batch of queries...I'll take a deeped llok into it next week and comment back

**Comment by monodo** [comment]
@ger-benjamin: after fetching the work done during the last 3 weeks, I get the following error when trying to start the dev server (make serve...). Any idea ?

```info ol Parsing dependencies
ERR! Unsatisfied dependency "ol.ext.pixelworks.Processor" in script: /home/sity/ngeo22/node_modules/openlayers/src/ol/source/raster.js 
Makefile:375: recipe for target '.build/contribs/gmf/examples/search.js' failed
make: *** [.build/contribs/gmf/examples/search.js] Error 1```

**Comment by monodo** [comment]
Actually, the same error is found on 2.2 branch

**Comment by ger-benjamin** [comment]
I've updated this PR with the last 2.2. It should work now.

**Comment by monodo** [comment]
Thanks, it's working now

**Pull request closed without merging**